pkg/applicationserver/device_cache.go

```diff
@@ -0,0 +1,63 @@
+// Copyright © 2020 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package applicationserver
+
+import (
+	"context"
+	"time"
+
+	"github.com/bluele/gcache"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/v3/pkg/unique"
+)
+
+// DeviceCache caches device protos.
+type DeviceCache interface {
+	Get(ctx context.Context, ids ttnpb.EndDeviceIdentifiers) (*ttnpb.EndDevice, error)
+	Set(ctx context.Context, ids ttnpb.EndDeviceIdentifiers, dev *ttnpb.EndDevice, ttl time.Duration) error
+}
+
+type memoryDeviceCache struct {
+	gcache.Cache
```

Comment:
Since we're in `pkg/applicationserver`, don't embed this type as that exposes all methods of `gcache.Cache` on `memoryDeviceCache`, leading to future errors in the future.